### DISCUSSION TITLE
UART BLE passthrough like HM-10 module

### DISCUSSION THREAD

**Discussion by DTSchneiderlein** [earlephilhower/arduino-pico#1469] [General]
I am looking for a way to use the RaspberryPI Pico W similar to a HM-10 module. Serial data that arrives via the UART port using the same UUID as the HM-10 module and vice versa. There is an Example doing this using normal Bluetooth, but how can I do with BLE?

**Comment by earlephilhower** [comment]
When I looked into this, it seemed to me that BLE does not have any concept of a serial port profile.  For HID devices like a keyboard or mouse, there is a standard way of making the BLE service visible, but serial ports are not HID devices.  You need to define special services and characteristics and then program your receiving host to access them.

Always happy to get an ArduinoBLE port for the core, of course. :)   The Arduino version of that library expects to control the entire BLE stack and send raw HCI commands to the chip, which is not supported by BTStack as I see it.

**Reply by DTSchneiderlein** [reply]
As far as I know, the HM-10 module uses a custom characteristic to send and receive the data it receives through the UART serial port. The UUID of the service is 0000ffe0-0000-1000-8000-00805f9b34fb and the characteristic UUID of the is 0000ffe1-0000-1000-8000-00805f9b34fb. When the HM-10 is supposed to send "23", it sets the value of the custom characteristic to "23" and then sends out a notification. So I think if I could define the same services and features as the HM-10 module and then pass the data from the serial port to the custom characteristic UUiD , it should work. But I can't find a way to do this. All the BLE examples are HID devices and because of the libraries I cannot change (or at least I did not get it right).  It would help me more to have an example that defines custom services and characteristics. Hope for your help :-)

**Reply by earlephilhower** [reply]
> It would help me more to have an example that defines custom services and characteristics.

You and me, both! 😆

For now, you need to use the BTStack `gatt` compiler and write naked BTstack code to make that happen using their examples (not BTStackLib, the raw BTStack in the Pico SDK).  In the future, hopefully someone steps up or I find a free weekend or two to hack something out...

**Reply by DTSchneiderlein** [reply]
Okay, I understand. So we have to wait for someone who is able to write a library for the Arduino IDE. I think with the Arduino BLE library it is possible to use custom UUIDs. Perhaps this could help this person to do so, just in the case this person is reading here :-))

**Reply by djpearman** [reply]
I've managed to get something like this to work. Here are code excerpts using a mixture of BTStackLib and BTstack:

```c++
#include <BTstackLib.h>
#include <ble/att_server.h>

uint8_t status[4];
uint16_t handleState;
UUID GATTServiceUUID = "4dcc3a1f-23b4-4921-b946-5809f15fd42e";
UUID GATTCharStatusUUID = "21be1adf-fe36-4faa-9d9d-3ac3dcc04019";

void setup() {
  BTstack.addGATTService(&GATTServiceUUID);
  handleState = BTstack.addGATTCharacteristic(&GATTCharStatusUUID, ATT_PROPERTY_READ | ATT_PROPERTY_NOTIFY, status, sizeof(status));
  BTstack.setup();
  BTstack.startAdvertising();
}

void loop() {
  BTstack.loop();
  att_server_notify(0x0040, handleState, status, sizeof(status));
}
```

This creates a BLE characteristic that is updated by the call to `att_server_notify()` and performs a notification. Note that the connection handle (0x0040) is hardcoded and taken from the debug output. Don't forget to adjust it.

Perhaps this is of some use to you. It's by no means complete as these lines are taken from a much larger codebase and is my first attempt at using BTstack. I'm still trying to get my head around how BTstack works, so any feedback or advice is very welcome :)

**Comment by lyusupov** [comment]
@DTSchneiderlein 
> way to use the RaspberryPI Pico W similar to a HM-10 module. Serial data that arrives via the UART port using the same UUID as the HM-10 module and vice versa

HM-10 compatible:
1) [Central role](https://github.com/lyusupov/SoftRF/blob/master/software/firmware/source/SkyView/BluetoothHelper.cpp#L1470)
2) [Peripheral role](https://github.com/lyusupov/SoftRF/blob/master/software/firmware/source/SoftRF/src/driver/Bluetooth.cpp#L2179)

Both implementations are based upon examples from [BTStack repository](https://github.com/bluekitchen/btstack/tree/master/example).


![](https://user-images.githubusercontent.com/5849637/235457545-b9a9135e-56b3-49b4-a6af-23438c21335e.png)

**Reply by DTSchneiderlein** [reply]
I tried to use your Bluetooth.cpp with the original LEPeripheral.ino and the original BTStackLib.h from the Pico Example. When I try to compile, I get "Compilation error: ../system/SoC.h: No such file or directory". Do need additional libraries?
